### ISSUE TITLE
Redirecting stdout/stderr without using "& FG or & BG constructs"

### ISSUE THREAD

**Issue by TMiguelT** [tomerfiliba/plumbum#354]
I love the ability to create pipelines and to do file redirections in Plumbum with the bash-like syntax, but I'm not a big fan of the `& FG`/`& BG` constructs for actually running the pipelines, because I feel that they're an unintuitive syntax. I feel that there should be an explicit method call to run these commands, and indeed my IDE treats these statements as doing nothing:
![screenshot from 2017-11-03 15-50-02](https://user-images.githubusercontent.com/5019367/32360953-eea2ed82-c0ae-11e7-8368-3fbd40a13504.png)

Is there a nice way to say "run this pipeline, with these redirections", without using these `&` constructs? I've looked at `run()` and `popen()`, but `run()` seems to return the entire stdout, which worries me because I'm doing large data processing and I don't want python to ever have to see this data. `popen()` is also more complex than I need, since you have to call `communicate`, and when you do so it still returns `stdout`, which I don't want.

**Comment by henryiii** [comment]
What IDE, just out of curiosity? I do admit they are not very pythonic, and they are very, very useful. I don't have time to work on this now, but might be able to look at it after 1.6.4. The code would be in the  `plumbum.commands.modifiers` file. FG should be easy, it just calls `cmd(retcode = self.retcode, stdin = None, stdout = None, stderr = None,
             timeout = self.timeout)`. BG is tricker, since it returns a Future, defined in the same file.

**Comment by TMiguelT** [comment]
It looks simple enough to make a new `command.fg()` and `command.bg()`method on the command class, move the `FG`/`BG` logic into there, and then get the `FG` and `BG` classes to use these methods. Is that an implementation that you would consider? I'm happy to make the changes, of course. Oh and that IDE is pycharm.

**Comment by henryiii** [comment]
Sure. The Future class might need to be moved.

Oh, I didn't recognize it, I guess that's not the theme I use. :)

**Comment by TMiguelT** [comment]
My only issue with this would be the interface. If I implemented `BaseCommand#bg` and `BaseCommand#fg`, they look somewhat similar to `BaseCommand#run` and `BaseCommand#bgrun`. The difference conceptually, is that these methods won't return any data to Python. Should they have a different name in that case?

**Comment by henryiii** [comment]
I think `fg` and `bg` are fine names, since we are already used to `FG` and `BG`. Unless there's an easy way to combine them into the run and bgrun methods.

**Comment by TMiguelT** [comment]
Okay fantastic, I've started thinking about the implementation. Does this look reasonable: https://github.com/TMiguelT/plumbum/commit/202fe17664b221d95e9410a5b58694a6e6ea6772

Sidenote, I tried to build the documentation to test it out, using `make html` in the docs directory, but that fails with:
```
Exception occurred:
  File "/home/michael/Programming/plumbum/plumbum/commands/modifiers.py", line 9, in <module>
    from plumbum.commands.base import AppendingStdoutRedirection, StdoutRedirection
ImportError: cannot import name 'AppendingStdoutRedirection'
```
Any idea why that is?

**Comment by henryiii** [comment]
I'll try to get back to you soon, swamped at the moment. Should have time soon.

**Comment by henryiii** [comment]
That looks pretty reasonable. Make a PR (add WIP if you think it needs more work), and make sure you have a test or two (though I can add that, but not quickly most likely).

I usually just let read-the-docs build the docs; I'm guessing that the problem you are seeing is because it's not finding plumbum. Did you install it first (into a virtual env if necessary)?

**Comment by henryiii** [comment]
Doesn't `plumbum.commands.modifier` import from `plumbum.commands.base`? I think this adds a circular import. We might need to move some things to a new file or make the import part of the apply method.

**Comment by TMiguelT** [comment]
I've tried to address these issues. The PR is here: https://github.com/tomerfiliba/plumbum/pull/386. However, I still can't get the docs to build, so I haven't really updated them (unless they auto-build from the docstrings, which I have updated)

**Comment by dkasak** [comment]
This should probably be closed now that #386 is merged.
